kratos/sources/model_part.cpp

```diff
@@ -1229,45 +1229,53 @@ void ModelPart::AddMasterSlaveConstraints(std::vector<IndexType> const& MasterSl
+                                                                                        ModelPart::DofsVectorType& rSlaveDofsVector,
+                                                                                        const ModelPart::MatrixType& RelationMatrix,
+                                                                                        const ModelPart::VectorType& ConstantVector,
+                                                                                        IndexType ThisIndex)
 {
 
     KRATOS_TRY
-    if (IsSubModelPart())
-    {
-        ModelPart::MasterSlaveConstraintType::Pointer p_new_constraint = mpParentModelPart->CreateNewMasterSlaveConstraint(ConstraintName, Id, rMasterDofsVector,
-                                                                                                                    rSlaveDofsVector,
-                                                                                                                    RelationMatrix,
-                                                                                                                    ConstantVector,
-                                                                                                                    ThisIndex);
-        GetMesh(ThisIndex).AddMasterSlaveConstraint(p_new_constraint);
-        GetMesh(ThisIndex).MasterSlaveConstraints().Unique();
-
-        return p_new_constraint;
-    }
+    MeshType& r_mesh = GetMesh(ThisIndex);
+    ModelPart::MasterSlaveConstraintType::Pointer p_new_constraint;
 
-    auto existing_constraint_iterator = GetMesh(ThisIndex).MasterSlaveConstraints().find(Id);
-    KRATOS_ERROR_IF(existing_constraint_iterator != GetMesh(ThisIndex).MasterSlaveConstraintsEnd() )
-        << "trying to construct an master-slave constraint with ID " << Id << " however a constraint with the same Id already exists";
-
-
-    //create the new element
-    ModelPart::MasterSlaveConstraintType const& r_clone_constraint = KratosComponents<MasterSlaveConstraintType>::Get(ConstraintName);
-    ModelPart::MasterSlaveConstraintType::Pointer p_new_constraint = r_clone_constraint.Create(Id, rMasterDofsVector,
-                                                                                        rSlaveDofsVector,
-                                                                                        RelationMatrix,
-                                                                                        ConstantVector);
+    if (IsSubModelPart()) {
+        // Defer constraint construction to the root model part
+        p_new_constraint = mpParentModelPart->CreateNewMasterSlaveConstraint(
+            ConstraintName,
+            Id,
+            rMasterDofsVector,
+            rSlaveDofsVector,
+            RelationMatrix,
+            ConstantVector,
+            ThisIndex);
+
+        // Add the constraint
+        if (&r_mesh != &mpParentModelPart->GetMesh(ThisIndex)) {
```

Comment:
you need to insert also in the current modelpart no?

Reply:
oh, i saw now the check you do first.

still, the mesh used in submodelparts should be normally different to the one used in the parent...so this should be a cornercase